src/cls/rbd/cls_rbd.cc

```diff
@@ -1988,6 +2054,52 @@ int snapshot_remove(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   return 0;
 }
 
+/**
+ * Moves a snapshot to the trash namespace.
+ *
+ * Input:
+ * @param snap_id the id of the snapshot to move to the trash (uint64_t)
+ *
+ * Output:
+ * @returns 0 on success, negative error code on failure
+ */
+int snapshot_trash_add(cls_method_context_t hctx, bufferlist *in,
+                       bufferlist *out)
+{
+  snapid_t snap_id;
+
+  try {
+    bufferlist::iterator iter = in->begin();
+    decode(snap_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  CLS_LOG(20, "snapshot_trash_add id=%" PRIu64, snap_id.val);
+
+  cls_rbd_snap snap;
+  std::string snapshot_key;
+  key_from_snap_id(snap_id, &snapshot_key);
+  int r = read_key(hctx, snapshot_key, &snap);
+  if (r == -ENOENT) {
+    return r;
+  }
+
+  if (snap.protection_status != RBD_PROTECTION_STATUS_UNPROTECTED) {
+    return -EBUSY;
+  } else if (cls::rbd::get_snap_namespace_type(snap.snapshot_namespace) ==
+               cls::rbd::SNAPSHOT_NAMESPACE_TYPE_TRASH) {
+    return -EINVAL;
```

Comment:
@dillaman Observing sporadic test failures due to snapshot_trash_add returning -EINVAL. In rbd_mirror_stress, when replaying 'snapshot remove' event [1]. In c librbd tests [2], in python tests [3].

[1] http://pulpito.ceph.com/trociny-2018-02-02_06:43:02-rbd-wip-mgolub-testing-distro-basic-smithi/
[2] http://pulpito.ceph.com/trociny-2018-02-02_06:43:52-rbd-wip-mgolub-testing-distro-basic-smithi/
[3] http://pulpito.ceph.com/trociny-2018-02-02_06:44:29-rbd-wip-mgolub-testing-distro-basic-smithi/

Reply:
Thanks -- I'll take a look.